frontends/concrete-python/concrete/fhe/mlir/context.py

```diff
@@ -238,29 +243,660 @@ def operation(
+                rhs of comparison
+
+            accept (Set[Comparison]):
+                set of accepted comparison outcomes
+
+        Returns:
+            Conversion:
+                result of comparison
+        """
+
+        if accept == {Comparison.GREATER}:  # pragma: no cover
+            return self.comparison(resulting_type, y, x, accept={Comparison.LESS})
+
+        if accept == {Comparison.GREATER, Comparison.EQUAL}:  # pragma: no cover
+            return self.comparison(resulting_type, y, x, accept={Comparison.LESS, Comparison.EQUAL})
+
+        assert accept in (
+            # equal
+            {Comparison.EQUAL},
+            # not equal
+            {Comparison.LESS, Comparison.GREATER},
+            # less and inverted greater
+            {Comparison.LESS},
+            # less equal and inverted greater equal
+            {Comparison.LESS, Comparison.EQUAL},
+        )
+
+        if x.bit_width > MAXIMUM_TLU_BIT_WIDTH or y.bit_width > MAXIMUM_TLU_BIT_WIDTH:
+            highlights = {
+                self.converting: [
+                    f"but only up to {MAXIMUM_TLU_BIT_WIDTH}-bit "
+                    f"comparison operations are supported"
+                ],
+            }
+            for operand in [x, y]:
+                if operand.bit_width > MAXIMUM_TLU_BIT_WIDTH:
+                    highlights[operand.origin] = [
+                        f"this {operand.bit_width}-bit value "
+                        f"is used as an operand to a comparison operation"
+                    ]
+                    if operand.bit_width != operand.original_bit_width:  # pragma: no cover
+                        highlights[operand.origin].append(
+                            "("
+                            f"note that it's assigned {operand.bit_width}-bits "
+                            f"during compilation because of its relation with other operations"
+                            ")"
+                        )
+            self.error(highlights)
+
+        assert resulting_type.is_encrypted and x.is_encrypted and y.is_encrypted
+
+        x_dtype = Integer(is_signed=x.is_signed, bit_width=x.original_bit_width)
+        y_dtype = Integer(is_signed=y.is_signed, bit_width=y.original_bit_width)
+
+        x_minus_y_min = x_dtype.min() - y_dtype.max()
+        x_minus_y_max = x_dtype.max() - y_dtype.min()
+
+        x_minus_y_range = [x_minus_y_min, x_minus_y_max]
+        x_minus_y_dtype = Integer.that_can_represent(x_minus_y_range)
+
+        maximum_input_bit_width = max(x.bit_width, y.bit_width)
+        if x_minus_y_dtype.bit_width <= maximum_input_bit_width:
+            return self.comparison_with_subtraction_trick(
+                resulting_type,
+                x,
+                y,
+                accept,
+                x_minus_y_dtype,
+            )
+
+        if x.original_bit_width != y.original_bit_width:
+            smaller = x if x.original_bit_width < y.original_bit_width else y
```

Comment:
all this logic should be part of self.comparison_with_clipping_trick
it could return None if it fails to apply and keep the logic internal